### ISSUE TITLE
dokan: FUSE for Windows

### ISSUE THREAD

**Issue by bradfitz** [perkeep/perkeep#146]
<pre><a href="http://dokan-dev.net/en/">http://dokan-dev.net/en/</a> is FUSE-like userspace filesystem for Windows.

There are C# and Ruby bindings, so Go probably is easy enough.

We should support it.</pre>

**Comment by glowingwire** [comment]
https://dokan-dev.github.io/

**Comment by bradfitz** [comment]
There is also now https://github.com/billziss-gh/winfsp

**Comment by Danilodum** [comment]
any guidance on setting up winfsp?

**Comment by octaviordz** [comment]
I've been exploring the FUSE for Windows compatibility and want to provide a summary of the information that I have gathered.

I initially came across a Go package with bindings for Dokan. While Dokan does support a FUSE API compatibility layer, the Go package I found, located at [this link](https://pkg.go.dev/github.com/keybase/client/go/kbfs/dokan), might not support it. What's even more critical is that the Go package used by Perkeep for FUSE access, "bazil.org/fuse," doesn't utilize the default FUSE API. Instead, it implements the binary protocol exchanged by file descriptors. This difference in implementation means that even if Windows Dokan-FUSE bindings existed, they would likely be of limited use because "bazil.org/fuse" operates at a different layer. For more information about "bazil.org/fuse," you can refer to this [presentation](https://bazil.org/talks/2013-06-10-la-gophers/#16).

It's important to note that the APIs of FUSE and Dokan are distinct. While I have found some information on how to map these APIs, such as [this link](https://github.com/dokan-dev/dokany/wiki/FUSE) and [this link](https://winfsp.dev/doc/SSHFS-Port-Case-Study/), it's crucial to recognize that "bazil.org/fuse" doesn't work at the FUSE API level. Unfortunately, I haven't been able to locate documentation about the FUSE protocol itself.

However, my more recent findings indicate that Windows offers native support for a feature similar to FUSE, known as the Projected File System. It can be accessed at [this link](https://learn.microsoft.com/en-us/windows/win32/projfs/projected-file-system). I should note that I haven't found any Go bindings for this feature.

Open questions arise from these discoveries:

- In the context of Perkeep, where should the Windows-FUSE compatibility layer be best implemented?
- Is a Windows-FUSE compatibility layer a worthwhile endeavor, or should Perkeep target Windows-specific APIs like Dokan or the Projected File System?

Your feedback and thoughts on these questions are greatly appreciated.

**Comment by tgulacsi** [comment]
If read-only would be enough, the not-so-new io/fs would work perfectly as a common target.
This could ease this porting:
  * have server implement io/fs (RO), or an extended form (RW, name it io/fs+)
  * have an io/fs+ -> bazil.org/fuse (or github.com/hanwen/go-fuse/v2, supports MacOS) exporter (this could be third party, as it would be useful for other projects, too)
  * have an io/fs+ -> Dokan (https://pkg.go.dev/github.com/binzume/dkango already has an io/fs.FS support!)
  
If this io.fs+ is close to what dkango uses, then that part is free.

**Comment by octaviordz** [comment]
The idea of implementing `io/fs+`, an extension of `io/fs` that includes write support appears to be a reasonable architecture solution.

Nevertheless base on empirical results from minimal testing of [dkango](https://pkg.go.dev/github.com/binzume/dkango) `io/fs+` implementation, it doesn't appear to be in a usable state. 
Upon running the example and attempting to create a new folder, the process fails with an error similar to the following:
![Error Screenshot](https://github.com/perkeep/perkeep/assets/348729/02738336-ceb0-4570-99ed-777285472b67)
Furthermore, when trying to delete a folder, another error is produced:
![Error Screenshot](https://github.com/perkeep/perkeep/assets/348729/895de5d9-ee0e-476d-9aa2-ec26922c0e24)

Back to `io/fs+` idea, you are implying that Perkeep should consider adopting such API/interface, has there been any work towards this idea. I'm not aware of the governance structure of this project, who can give the "approved in principle" status to an idea.

**Comment by tgulacsi** [comment]
Yes, the common fs layer came from when I tried to implement a WebDAV server. It has cried to not reimplement the wheel, but the existing fuse api did not bend into shape (was too complex).

**Comment by djdv** [comment]
I can give an anecdote FWIW.
Prior to `fs.FS` I had a project that defined an interface that looked very similar, and used it to integrate with host APIs like FUSE, 9P, and others. That code would later migrate to `fs.FS` and utilize the extension pattern demonstrated by Russ in some of his presentations on it.

The implementation for handling `fs.FS` -> FUSE utilized the [cgo-fuse](https://github.com/winfsp/cgofuse) libray. Which uses WinFSP underneath for Windows/NT, and whatever libfuse implementation for other POSIX systems.

I'm primarily a Windows user and do development on a Windows system so I can vouch for both that library and the pattern of using `fs.FS` with extensions to integrate with it.
It hasn't given me any notable issues during development or use, and has made it nice to have a single implementation that supports WinFSP+FUSE on an increasing list of platforms.

If any of this sounds interesting I can provide more details on these implementations that might apply to Perkeep's future implementation as well. However, I'm currently away and won't be back until ~the 6th.
